File: mini-claw-code-starter-py/src/mini_claw_code_starter_py/tools/bash.py
from __future__ import annotations

from typing import Any

from ..types import ToolDefinition


class BashTool:
    def __init__(self) -> None:
        self._definition = None

    @property
    def definition(self) -> ToolDefinition:
        if self._definition is None:
            raise RuntimeError("BashTool.new() must initialize self._definition")
        return self._definition

    @classmethod
    def new(cls) -> "BashTool":
        raise NotImplementedError(
            "Define a bash tool with one required 'command' string parameter"
        )

    async def call(self, args: Any) -> str:
        raise NotImplementedError(
            "Run bash -lc <command>, combine stdout and stderr, and return '(no output)' when empty"
        )

File: mini-claw-code-starter-ts/src/tools/bash.ts
import { spawn } from "node:child_process";

import type { JsonValue, Tool } from "../types";
import { ToolDefinition } from "../types";

export class BashTool implements Tool {
  readonly toolDefinition: ToolDefinition;

  constructor() {
    this.toolDefinition = ToolDefinition.new(
      "bash",
      "Run a bash command and return its output.",
    ).param("command", "string", "The bash command to run", true);
  }

  static new(): BashTool {
    return new BashTool();
  }

  definition(): ToolDefinition {
    return this.toolDefinition;
  }

  async call(_args: JsonValue): Promise<string> {
    void spawn;
    throw new Error(
      "TODO(ch4): run `bash -lc`, collect stdout/stderr, and return a combined string",
    );
  }
}
